hw/ip/prim/dv/prim_ram_scr/cpp/scramble_model.cc

```diff
@@ -0,0 +1,373 @@
+  }
+
+  return out;
+}
+
+// Apply a full set of subsitution/permutation rounds for encrypt to the
+// incoming byte vector
+static std::vector<uint8_t> scramble_subst_perm_enc(
+    const std::vector<uint8_t> &in, const std::vector<uint8_t> &key,
+    uint32_t bit_width, uint32_t num_rounds) {
+  assert(in.size() >= ((bit_width + 7) / 8));
+  assert(key.size() >= ((bit_width + 7) / 8));
+  std::vector<uint8_t> state(in);
+
+  for (int i = 0; i < num_rounds; ++i) {
+    for (int j = 0; j < in.size(); ++j) {
+      state[j] = state[j] ^ key[j];
+    }
+
+    state = scramble_sbox_layer(state, bit_width, false);
+    state = scramble_flip_layer(state, bit_width);
+    state = scramble_perm_layer(state, bit_width, false);
+  }
+
+  for (int i = 0; i < in.size(); ++i) {
+    state[i] = state[i] ^ key[i];
+  }
+
+  return state;
+}
+
+// Apply a full set of substitution/permutation rounds for decrypt to the
+// incoming byte vector
+static std::vector<uint8_t> scramble_subst_perm_dec(
+    const std::vector<uint8_t> &in, const std::vector<uint8_t> &key,
+    uint32_t bit_width, uint32_t num_rounds) {
+  assert(in.size() >= ((bit_width + 7) / 8));
+  assert(key.size() >= ((bit_width + 7) / 8));
+
+  std::vector<uint8_t> state(in);
+
+  for (int i = 0; i < num_rounds; ++i) {
+    for (int j = 0; j < in.size(); ++j) {
+      state[j] = state[j] ^ key[j];
+    }
+
+    state = scramble_perm_layer(state, bit_width, true);
+    state = scramble_flip_layer(state, bit_width);
+    state = scramble_sbox_layer(state, bit_width, true);
+  }
+
+  for (int i = 0; i < in.size(); ++i) {
+    state[i] = state[i] ^ key[i];
+  }
+
+  return state;
+}
+
+// Generate a keystream for XORing with data using PRINCE. Multiple PRINCE
+// instantiatons are used when the keystream is greater than a single PRINCE
+// width (64-bit).
+static std::vector<uint8_t> scramble_gen_keystream(
+    const std::vector<uint8_t> &addr, uint32_t addr_width,
+    const std::vector<uint8_t> &nonce, const std::vector<uint8_t> &key,
+    uint32_t keystream_width, uint32_t num_half_rounds) {
+  assert(key.size() == (kPrinceWidthByte * 2));
+
+  // Determine how many PRINCE replications are required
+  uint32_t num_prince = (keystream_width + kPrinceWidth - 1) / kPrinceWidth;
+
+  std::vector<uint8_t> keystream;
+
+  for (int i = 0; i < num_prince; ++i) {
+    // Initial vector is data for PRINCE to encrypt. Formed from nonce and data
+    // address
+    std::vector<uint8_t> iv(8, 0);
+
+    for (int j = 0; j < kPrinceWidth; ++j) {
+      if (j < addr_width) {
+        // Bottom addr_width bits of IV are address
+        or_vector_bit(iv, j, read_vector_bit(addr, j));
+      } else {
+        // Other bits are taken from nonce. Each PRINCE instaniations will used
+        // different nonce bits.
+        or_vector_bit(
+            iv, j,
+            read_vector_bit(
+                nonce, (j - addr_width) + i * (kPrinceWidth - addr_width)));
```

Comment:
The auto-formatter has made a bit of a meal of this. Maybe worth putting the source bit index calculation into a helper variable (then I think this will end up on just 2 lines).